module-sys/SystemManager/SystemManagerCommon.cpp

```diff
@@ -444,20 +463,41 @@ namespace sys
         return "none";
     }
 
-    void SystemManagerCommon::preCloseRoutine(CloseReason closeReason)
+    void SystemManagerCommon::InitiateSystemCloseSequence(CloseReason closeReason)
     {
         for (const auto &service : servicesList) {
-            auto msg = std::make_shared<ServiceCloseReasonMessage>(closeReason);
-            bus.sendUnicast(std::move(msg), service->GetName());
-            readyForCloseRegister.push_back(service->GetName());
+            servicesToClose.push_back(service->GetName());
         }
-
-        // stored to be used later in CloseServices
         this->closeReason = closeReason;
 
-        servicesPreShutdownRoutineTimeout = sys::TimerFactory::createPeriodicTimer(
-            this, "servicesPreShutdownRoutine", preShutdownRoutineTimeout, [this](sys::Timer &) { CloseServices(); });
-        servicesPreShutdownRoutineTimeout.start();
+        serviceCloseTimer = sys::TimerFactory::createSingleShotTimer(
+            this, "serviceCloseTimer", serviceCloseResponseTimeout, [this](sys::Timer &) { CloseServices(); });
+        serviceCloseTimer.start();
+
+        /// Start the sequence by sending close request to the first service on the list
+        bus.sendUnicast(std::make_shared<ServiceCloseReasonMessage>(closeReason), servicesToClose.front());
```

Comment:
Is it possible that some service doesn't close at all and blocks? If that happens CloseServices() is called, forcefully closing all the services left. When closing services in sequence, shouldn't CloseServices() close particular stuck service instead? Or is it out of the scope here?

Reply:
Yes, such a situation can happen. We can discuss it as I'm not sure if closing only non-responding service and then proceed with the rest normally will make any difference. In either case, it's abnormal situation that cannot be handled properly in system without MMU. Anyway I'm open for discussion.

Reply:
Fair point. I guess, it is acceptable to "abort" the proper termination anyway.